google-cloud-storage/.toys/benchmark/perform.rb

```diff
@@ -0,0 +1,213 @@
+
+HEADER = [
+  "Op",
+  "ObjectSize",
+  "AppBufferSize",
+  "LibBufferSize",
+  "Crc32cEnabled",
+  "MD5Enabled",
+  "ApiName",
+  "ElapsedTimeUs",
+  "CpuTimeUs",
+  "Status",
+  "RunID"
+]
+
+desc "Run performance benchmark"
+
+flag :min_size, "--min-size=EVENT" do |f|
+  f.default DEFAULT_MIN_SIZE
+  f.accept Integer
+  f.desc "Minimum object size in bytes."
+end
+flag :max_size, "--max-size=PATH" do |f|
+  f.default DEFAULT_MAX_SIZE
+  f.accept Integer
+  f.desc "Maximum object size in bytes."
+end
+flag :num_samples, "--num_samples=COMMIT" do |f|
+  f.default DEFAULT_NUM_SAMPLES
+  f.accept Integer
+  f.desc "Number of iterations."
+end
+flag :bucket_location, "--r=NAMES" do |f|
+  f.accept Array
+  f.default DEFAULT_BUCKET_LOCATION
+  f.desc "Bucket location."
+end
+flag :output, "--o[=FILES]" do |f|
+  f.accept String
+  f.default "benchmarking#{TIMESTAMP}.csv"
+  f.desc "File output results to."
+end
+flag :bucket_name, "--b=NAME" do |f|
+  f.accept String
+  f.default "benchmarking#{TIMESTAMP}"
+  f.desc "Storage bucket name."
+end
+
+include :terminal, styled: true
+include :bundler
+include :fileutils
+
+def run
+  require "google/cloud/storage"
+  require "securerandom"
+  require "csv"
+  
+  # Create a storage bucket to run benchmarking
+  storage = Google::Cloud::Storage.new
+  @bucket = storage.bucket(bucket_name) rescue nil
+  @bucket = storage.create_bucket(bucket_name, location: bucket_location) if @bucket.nil?
+
+  @results = []
+  (1..num_samples).each do |i|
+    puts "Running W1R3 iteration number #{i}", :bold, :yellow
+    w1r3_benchmark_runner
+  end
+
+  # Output to CSV file
+  output_to_csv
+  puts "Succesfully ran benchmarking. Please find your output log at #{output}", :bold, :cyan
+
+  cleanup_resources
+end
+
+def w1r3_benchmark_runner
+  puts "Generate Write-1-Read-3 workload.", :bold
+  # generate randmon size in bytes using a uniform distribution
+  size = rand(min_size..max_size)
+  object_name = "#{TIMESTAMP}-#{SecureRandom.uuid}"
+
+  # generate random checksumming type: md5, crc32c or None
+  checksum = CHECKSUM.sample
+
+  @results << log_performance("WRITE", size, checksum, write(object_name, size, checksum))
+
+  (1..3).each do |i|
+    @results << log_performance("READ[#{i}]", size, checksum, read(object_name, checksum))
+  end
+end
+
+def write object_name, size, checksum
+  puts "Perform an upload and return latency.", :bold, :blue
+  object = @bucket.file object_name
+  file_path = "#{Dir.getwd}/#{SecureRandom.uuid}"
+
+  # Create random file locally on disk
+  file = File.new(file_path, "wb+")
+  file.write(SecureRandom.random_bytes(size))
+  file.rewind
+
+  start_time = get_time_in_ms
+  @bucket.create_file file, object_name, checksum: checksum, if_generation_match: 0
+  end_time = get_time_in_ms
+
+  # convert nanoseconds to microseconds
+  elapsed_time = (end_time - start_time).round
+
+  # Clean up local file
+  cleanup_file file_path
+  elapsed_time
+rescue Exception => e
+  NOT_SUPPORTED
```

Comment:
Forgive me for my ignorance on Ruby here, can you explain what happens here? Does this execute in the case that there is an error in the upload?

Reply:
In case there's an error/exception, we just return -1 for elapsed time. Is the expectation to exit the flow instead?

Reply:
No, this is fine.